TestLoopbackHostPort fails on non-ipv6 machine

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


**What happened**:
```
root@valving1:~/k8s_ws/src/k8s.io/kubernetes# make check WHAT=./vendor/k8s.io/apiserver/pkg/server GOFLAGS=-v KUBE_TEST_ARGS='-run ^TestLoopbackHostPort$'
Running tests for APIVersion: v1,admissionregistration.k8s.io/v1beta1,admission.k8s.io/v1beta1,apps/v1,apps/v1beta1,apps/v1beta2,auditregistration.k8s.io/v1alpha1,authentication.k8s.io/v1,authentication.k8s.io/v1beta1,authorization.k8s.io/v1,authorization.k8s.io/v1beta1,autoscaling/v1,autoscaling/v2beta1,autoscaling/v2beta2,batch/v1,batch/v1beta1,batch/v2alpha1,certificates.k8s.io/v1beta1,coordination.k8s.io/v1beta1,coordination.k8s.io/v1,extensions/v1beta1,events.k8s.io/v1beta1,imagepolicy.k8s.io/v1alpha1,networking.k8s.io/v1,networking.k8s.io/v1beta1,node.k8s.io/v1alpha1,node.k8s.io/v1beta1,policy/v1beta1,rbac.authorization.k8s.io/v1,rbac.authorization.k8s.io/v1beta1,rbac.authorization.k8s.io/v1alpha1,scheduling.k8s.io/v1alpha1,scheduling.k8s.io/v1beta1,scheduling.k8s.io/v1,settings.k8s.io/v1alpha1,storage.k8s.io/v1beta1,storage.k8s.io/v1,storage.k8s.io/v1alpha1,
+++ [0410 07:58:43] Running tests without code coverage
=== RUN   TestLoopbackHostPort
--- FAIL: TestLoopbackHostPort (0.00s)
    config_selfclient_test.go:63: expected IPv6 host to be loopback, got "localhost"
FAIL
FAIL k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server 0.118s
Makefile:184: recipe for target 'check' failed
make: *** [check] Error 1
```

LoopbackHostPort code is calling the [net.InterfaceAddrs()](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/config_selfclient.go#L84) which depends on the machine on which this test is running. Especially in this particular testase it looks for the ipv6 ip address which does not exist because the machine is not configured for ipv6.

Because of the LoopbackHostPort function bahaves based on the interfaces configured, test will fail if machine is not really configured with ipv6. Need to rewrite the testcase to handle such situation.

**What you expected to happen**:
Testcase should pass

--- a/staging/src/k8s.io/apiserver/pkg/server/config_selfclient_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/config_selfclient_test.go
@@ -21,7 +21,15 @@ import (
 	"testing"
 )
 
-func TestLoopbackHostPort(t *testing.T) {
+func TestLoopbackHostPortIPv4(t *testing.T) {
+	_, ipv6only, err := isIPv6LoopbackSupported()
+	if err != nil {
+		t.Fatalf("fail to enumerate network interface, %s", err)
+	}
+	if ipv6only {
+		t.Fatalf("no ipv4 loopback interface")
+	}
+
 	host, port, err := LoopbackHostPort("1.2.3.4:443")
 	if err != nil {
 		t.Fatalf("unexpected error: %v", err)
@@ -43,8 +51,17 @@ func TestLoopbackHostPort(t *testing.T) {
 	if port != "443" {
 		t.Fatalf("expected 443 as port, got %q", port)
 	}
+}
+func TestLoopbackHostPortIPv6(t *testing.T) {
+	ipv6, _, err := isIPv6LoopbackSupported()
+	if err != nil {
+		t.Fatalf("fail to enumerate network interface, %s", err)
+	}
+	if !ipv6 {
+		t.Fatalf("no ipv6 loopback interface")
+	}
 
-	host, port, err = LoopbackHostPort("[ff06:0:0:0:0:0:0:c3]:443")
+	host, port, err := LoopbackHostPort("[ff06:0:0:0:0:0:0:c3]:443")
 	if err != nil {
 		t.Fatalf("unexpected error: %v", err)
 	}
@@ -62,8 +79,28 @@ func TestLoopbackHostPort(t *testing.T) {
 	if ip := net.ParseIP(host); ip == nil || !ip.IsLoopback() || ip.To4() != nil {
 		t.Fatalf("expected IPv6 host to be loopback, got %q", host)
 	}
-
 	if port != "443" {
 		t.Fatalf("expected 443 as port, got %q", port)
 	}
+}
+
+func isIPv6LoopbackSupported() (ipv6 bool, ipv6only bool, err error) {
+	addrs, err := net.InterfaceAddrs()
+	if err != nil {
+		return false, false, err
+	}
+	ipv4 := false
+	for _, address := range addrs {
+		ipnet, ok := address.(*net.IPNet)
+		if !ok || !ipnet.IP.IsLoopback() {
+			continue
+		}
+		if ipnet.IP.To4() == nil {
+			ipv6 = true
+			continue
+		}
+		ipv4 = true
+	}
+	ipv6only = ipv6 && !ipv4
+	return ipv6, ipv6only, nil
 }

